BibBuddy/app/src/main/java/de/bibbuddy/BookFragment.java

```diff
@@ -108,6 +125,213 @@ public boolean onOptionsItemSelected(MenuItem item) {
+    reqAlertDialog.setPositiveButton(R.string.ok,
+        (dialog, which) -> ActivityCompat.requestPermissions(getActivity(),
+            new String[] {Manifest.permission.WRITE_EXTERNAL_STORAGE},
+            STORAGE_PERMISSION_CODE));
+    reqAlertDialog.setNegativeButton(R.string.storage_permission_cancel_btn,
+        (dialog, which) -> dialog.dismiss());
+    reqAlertDialog.create().show();
+  }
+
+  /**
+   * Callback method, that checks the result from requesting permissions.
+   *
+   * @param requestCode unique integer value for the requested permission
+   *                    This value is given by the programmer.
+   * @param permissions array of requested name(s)
+   *                    of the permission(s)
+   * @param grantResults grant results for the corresponding permissions
+   *                     which is either PackageManager.PERMISSION_GRANTED
+   *                     or PackageManager.PERMISSION_DENIED.
+   */
+  @Override
+  public void onRequestPermissionsResult(int requestCode,
+                                         String[] permissions, int[] grantResults) {
+
+    switch (requestCode) {
+
+      case STORAGE_PERMISSION_CODE:
+        if (grantResults.length > 0
+            && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
+          createBibFile("Download", shelfName);
+          retrieveBibContent();
+          writeBibFile("Download", shelfName, bibExportContent);
+        } else {
+          Toast.makeText(getContext(), R.string.storage_permission_denied,
+              Toast.LENGTH_SHORT).show();
+        }
+        break;
+
+      default:
+    }
+  }
+
+  private void createBibFile(String folderName, String fileName) {
+    try {
+      String rootPath = Environment.getExternalStorageDirectory() + "/" + folderName + "/";
+      File root = new File(rootPath);
+
+      if (!root.exists()) {
+        root.mkdirs();
+      }
+
+      File file = new File(rootPath + fileName + ".bib");
+      if (file.exists()) {
+        file.delete();
+      }
+      file.createNewFile();
+
+      FileOutputStream out = new FileOutputStream(file);
+      out.flush();
+      out.close();
+
+      Toast.makeText(getContext(), getString(R.string.exported_file_stored_in) + '\n'
+          + "/" + folderName + "/" + fileName + ".bib", Toast.LENGTH_LONG).show();
+
+    } catch (Exception e) {
+      e.printStackTrace();
+    }
+  }
+
+  private void retrieveBibContent() {
+
+    DatabaseHelper dbHelper = new DatabaseHelper(getContext());
+    BookDao bd = new BookDao(dbHelper);
+    NoteDao nd = new NoteDao(dbHelper);
+
+    List<Long> bookIdsCurrShelf = bd.getAllBookIdsForShelf(shelfId);
+    List<Author> authorsCurrBook = new ArrayList<>();
+    List<Long> notesCurrBook = new ArrayList<>();
+    String allNotesCurrBook = "";
+    String authorNamesCurrBook = "";
+
+    //for each book in the current shelf
+    for (int i = 0; i < bookIdsCurrShelf.size(); i++) {
+      Book currBook;
```

Comment:
Divided initialization & declaration can be resolved here. See next three changes, too!

```suggestion
      Long currBookId = bookIdsCurrShelf.get(i);
```